File: backend/app/models/bird.py
"""Pydantic models for bird data."""

from pydantic import BaseModel, Field


class BirdImage(BaseModel):
    """A single image associated with a bird species."""

    url: str
    source: str = "inaturalist"  # "inaturalist" | "wikimedia"
    license: str = ""
    attribution: str = ""
    quality: str = "high"  # "high" | "medium" | "low"
    is_primary: bool = True


class Bird(BaseModel):
    """A bird species document stored in Cosmos DB."""

    id: str  # speciesCode (partition-independent unique id)
    species_code: str
    sci_name: str
    com_name: str
    family_code: str  # partition key
    family_com_name: str
    order: str = ""
    sort_order: int = 0
    inat_taxon_id: int | None = None
    images: list[BirdImage] = Field(default_factory=list)
    audio_url: str = ""
    audio_attribution: str = ""
    wikipedia_url: str = ""
    lookalikes: list[str] = Field(default_factory=list)
    data_version: str = ""

    @property
    def primary_image_url(self) -> str:
        """URL of the primary image, or empty string if none."""
        for img in self.images:
            if img.is_primary:
                return img.url
        return self.images[0].url if self.images else ""


class BirdSummary(BaseModel):
    """Lightweight bird representation for list endpoints."""

    id: str
    species_code: str
    sci_name: str
    com_name: str
    family_code: str
    family_com_name: str
    image_url: str = ""


class ChatRequest(BaseModel):
    """Request body for the chat proxy endpoint."""

    messages: list[dict]


class ChatResponse(BaseModel):
    """Response from the chat proxy endpoint."""

    role: str
    content: str


class DataVersion(BaseModel):
    """Current dataset version info."""

    version: str
    total_species: int
    image_coverage_pct: float

File: src/types/bird.ts
/**
 * Bird types — mirrors the backend Pydantic models.
 * These are the canonical types consumed throughout the app.
 */

export interface BirdImage {
    url: string;
    source: string;
    license: string;
    attribution: string;
    quality: string;
    is_primary: boolean;
}

/** Full bird document from Cosmos DB */
export interface Bird {
    id: string;
    species_code: string;
    sci_name: string;
    com_name: string;
    family_code: string;
    family_com_name: string;
    order: string;
    sort_order: number;
    inat_taxon_id: number | null;
    images: BirdImage[];
    audio_url: string;
    audio_attribution: string;
    wikipedia_url: string;
    lookalikes: string[];
    /** Fraction of eBird checklists worldwide reporting this species (0–1). */
    global_frequency: number;
    data_version: string;
}

/** Lightweight summary for list views */
export interface BirdSummary {
    id: string;
    species_code: string;
    sci_name: string;
    com_name: string;
    family_code: string;
    family_com_name: string;
    image_url: string;
    wikipedia_url?: string;
    global_frequency: number;
    lookalike_count: number;
}

/** Summary with all image URLs for lookalike mode */
export interface LookalikeBirdSummary extends BirdSummary {
    image_urls: string[];
}

export interface BirdFamily {
    family_code: string;
    family_com_name: string;
}

/** An eBird geographic region (country, state, or county). */
export interface Region {
    code: string;
    name: string;
}

/** Helper: get the primary image URL from a Bird */
export function getPrimaryImageUrl(bird: Bird | BirdSummary): string {
    if ("image_url" in bird) return bird.image_url;
    const primary = bird.images.find((img) => img.is_primary);
    return primary?.url ?? bird.images[0]?.url ?? "";
}
